### PR TITLE
Add "broken" argument to test() and revise should_fail usage.

### PR DESCRIPTION
**Pull request by trhd**
There are some issues with how executed tests are currently reported to the user. Pull request #4466 introduced the "expected fail" and "expected pass" test "categories" that come to play when the `should_fail` argument is given to `test()`. Unfortunately this doesn't really work as, unless I'm horribly mistaken, the `should_fail`'s usage up to that point had not been (only) to indicate broken/incomplete tests but rather just to reverse the outcome of the test from the executed program's return value.

I see the following issues with the current situation:

- The "expected fail" tests are not necessarily failing and should not be reported as such (e.g. a test can be made to ensure that a binary returns an error with invalid input). A non-zero exit code from the executed program doesn't mean that *test* has failed.

- An "unexpected pass" (a test that `should_fail` but didn't) doesn't necessarily indicate that something has been fixed but rather that something has broken.

- Tests that pass "unexpectedly" (i.e. `should_fail` but didn't) aren't considered as errors (by design if I have understood correctly) and the overall exit status of the executed tests is thus unaffected by them. This means that a test that `should_fail` can become broken and be completely missed in CI since its not a "real error".

If something like the XPASS/XFAIL feature is really needed (#3296) I believe it needs to be possible to differentiate them from the "regular" `should_fail` case. I thought about adding a keyword `expect_failure` to the `test()` function but as it would be so close to the `should_fail` argument I went with `broken` instead to make the difference to the `should_fail` argument more obvious. The `should_fail` argument could of course be renamed as well but that would break backwards compatibility.

TL;DR Implement XPASS/XFAIL feature with a new `broken` argument to `test()`.

And also, stop counting timed out tests in both timedout tests and failing tests in the test summary.

### CODE REVIEW

**Comment by jpakkane** [comment]
Master has been fixed. Please rebase so CI gets run. Thanks.

**Comment by trhd** [comment]
"Rebased" the changes as there had been quite many changes to the relevant testing code...

Three failures in the checks above:

- 2x vs2017: Some linker error `LINK : fatal error LNK1104: cannot open file 'libphobos2.so'`. Seems to be breaking ~all pull requests at the moment.
- cygwin: A broken test case that was now revealed when the exit code for the failing (i.e. passing) `should_fail` case was fixed. If you read the test's logs from some other pull request you'll see the test has a UNEXPECTEDPASS result.

**Comment by nirbheek** [comment]
CI has been fixed in master, please rebase.

**Comment by trhd** [comment]
The cygwin issue described above remains...

**Comment by jpakkane** [comment]
> Tests that pass "unexpectedly" (i.e. should_fail but didn't) aren't considered as errors (by design if I have understood correctly) 

AFAIUI this is the real issue here. If a should fail test passes, it should be marked as an error. What makes you think that this is by design?

**Comment by trhd** [comment]
> > Tests that pass "unexpectedly" (i.e. should_fail but didn't) aren't considered as errors (by design if I have understood correctly)
> 
> AFAIUI this is the real issue here. If a should fail test passes, it should be marked as an error. What makes you think that this is by design?

Just the impression I got from [here](https://www.gnu.org/software/automake/manual/html_node/Generalities-about-Testing.html) and [here](https://docs.pytest.org/en/latest/skipping.html#imperative-xfail-from-within-a-test-or-setup-function). Also, if we think about broken/incomplete tests in general, I wouldn't say that it its always the case that they always fail or pass but instead they might be flaky and pass/fail occasionally. Treating xpass as a failure in those situations might be problematic.

I also checked how pytests treats xpass and it would seem that it treats it as a non-error:

```
$ cat pytest_test.py
import pytest
@pytest.mark.xfail
def test_xpass():
    pass
$ pytest pytest_test.py
============================= test session starts ==============================
platform linux -- Python 3.7.2, pytest-4.3.0, py-1.7.0, pluggy-0.8.1
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/trhd/.hypothesis/examples')
rootdir: /home/trhd, inifile:
plugins: subtesthack-0.1.1, localserver-0.5.0, hypothesis-4.7.1
collected 1 item

pytest_test.py X                                                         [100%]

========================== 1 xpassed in 0.01 seconds ===========================
$ echo $?
0
```

**Comment by trhd** [comment]
One possibility would be to just drop the whole xpass/xfail approach completely and let `should_fail` just reverse the result like it used to but as those were specifically requested and implemented by others I suppose there is some demand for those.

**Comment by trhd** [comment]
Added a change to fix the broken test case. Not really convinced that I've now comprehended all the related aspects of library symbols on cygwin, but these changes would seem to fix the test at least.

**Comment by jpakkane** [comment]
Adding a new type for this seems unnecessary. I wrote a simpler alternative approach in #5289. Does it do all the things you need?

**Comment by trhd** [comment]
I'd say it *does* all the things that I need, but I do think that some semantic issues would remain. Mainly, `should_fail` is documented and behaving like "just inverse the executable's outcome" (i.e. nothing about xpass/xfail) but when actually running the tests we are using the "expected fail" and "unexpected pass" nomenclature.

To get me completely happy with all this I'd say either the xpass/xfail use cases should actually be supported or the terminology referring to those should be dropped. I see no added value reporting the tests whose executable's return value has been inverted as "expected fail". Qite the opposite actually. The below for example is from a CI run where there was no broken/incomplete tests but everything went as designed:

```
[snip]
311/316 optargs:subarguments / Invalid subargument causes an error.  EXPECTEDFAIL 0.37 s 
312/316 optargs:subarguments / Acceptable arguments are accepted.  OK       0.37 s 
313/316 optargs:subarguments / Optional subargument is accepted.  OK       0.37 s 
314/316 optargs:subarguments / Optional "any" subsubargument is accepted.  OK       0.37 s 
315/316 optargs:subarguments / Optional "fixed" subsubargument is accepted.  OK       0.37 s 
316/316 optargs:subarguments / Invalid subsubargument causes an error.  EXPECTEDFAIL 0.37 s 

Ok:                  252
Expected Fail:        64
Fail:                  0
Unexpected Pass:       0
Skipped:               0
Timeout:               0
```

At least I would in these cases much prefer to see something like this:

```
[snip]
311/316 optargs:subarguments / Invalid subargument causes an error.  PASS     0.02 s 
312/316 optargs:subarguments / Acceptable arguments are accepted.  PASS     0.01 s 
313/316 optargs:subarguments / Optional subargument is accepted.  PASS     0.01 s 
314/316 optargs:subarguments / Optional "any" subsubargument is accepted.  PASS     0.01 s 
315/316 optargs:subarguments / Optional "fixed" subsubargument is accepted.  PASS     0.01 s 
316/316 optargs:subarguments / Invalid subsubargument causes an error.  PASS     0.01 s 

Pass:      316
Fail:        0
Skip:        0
Timeout:     0
```

All this being said, if I'm the only one being irritated by all this we can definitely just merge #5289 (and fix the broken test) and call it a day.

**Comment by jpakkane** [comment]
There are two different issues here:

- whether unexpected pass should count as an error
- how should messages be printed

The former we agree on (yes it should).

The latter is trickier. I agree that printing anything with the text "FAIL" on success is a bit bad, but on the other hand there is value in separating xfails from regular passes. What about printing PASS on the individual test lines but keeping the total summary as is? So something like:

```
316/316 optargs:subarguments / Invalid subsubargument causes an error.  PASS 0.37 s 

Ok:                  252
Expected Fail:        64
Fail:                  0
Unexpected Pass:       0
Skipped:               0
Timeout:               0
```

**Comment by trhd** [comment]
I think that would be bad cause then you wouldn't know which tests caused the "Expected fail" counter to increase (assuming one would care).

Perhaps ca14355 is however good enough for now. If this topic is brought up again later (by someone other than me) I can try to pitch in.

**Comment by jpakkane** [comment]
I merged the code changes from the other branch. I'll close this, we can do the renaming with proper deep thought if still comes up.
